### ISSUE TITLE
Error when adding SSH Keys: "illegal base64 data at input byte 36"

### ISSUE THREAD

**Issue by Svedrin** [ansible-semaphore/semaphore#1132]
I have installed Ansible Semaphore using the `semaphoreui/semaphore:latest` Docker image (currently, this is `v2.8.77`). It seems to work fine, but when I try to add an SSH key through the UI, I'm only getting a red "Request failed with status code 400" error and the console shows this traceback:

```
time="2022-12-30T10:15:43Z" level=error msg="illegal base64 data at input byte 36"
goroutine 632 [running]:
runtime/debug.Stack()
        /usr/local/go/src/runtime/debug/stack.go:24 +0x65
runtime/debug.PrintStack()
        /usr/local/go/src/runtime/debug/stack.go:16 +0x19
github.com/ansible-semaphore/semaphore/api/helpers.WriteError({0x1fc4330, 0xc00006d500}, {0x1fc1440?, 0x22b9840?})
        /go/src/github.com/ansible-semaphore/semaphore/api/helpers/helpers.go:92 +0x1ae
github.com/ansible-semaphore/semaphore/api/projects.AddKey({0x1fc4330, 0xc00006d500}, 0xc000c3f478?)
        /go/src/github.com/ansible-semaphore/semaphore/api/projects/keys.go:91 +0x66d
net/http.HandlerFunc.ServeHTTP(0xc0008d2a00?, {0x1fc4330?, 0xc00006d500?}, 0x9a00c0?)
        /usr/local/go/src/net/http/server.go:2084 +0x2f
github.com/ansible-semaphore/semaphore/api/projects.ProjectMiddleware.func1({0x1fc4330, 0xc00006d500}, 0x0?)
        /go/src/github.com/ansible-semaphore/semaphore/api/projects/project.go:41 +0x3bb
net/http.HandlerFunc.ServeHTTP(0x1fc4330?, {0x1fc4330?, 0xc00006d500?}, 0xc000051001?)
        /usr/local/go/src/net/http/server.go:2084 +0x2f
github.com/ansible-semaphore/semaphore/api.authentication.func1({0x1fc4330, 0xc00006d500}, 0xc0004bb600?)
        /go/src/github.com/ansible-semaphore/semaphore/api/auth.go:109 +0x59
net/http.HandlerFunc.ServeHTTP(0x98fec0?, {0x1fc4330?, 0xc00006d500?}, 0xc?)
        /usr/local/go/src/net/http/server.go:2084 +0x2f
github.com/ansible-semaphore/semaphore/api.JSONMiddleware.func1({0x1fc4330, 0xc00006d500}, 0x1b?)
        /go/src/github.com/ansible-semaphore/semaphore/api/router.go:35 +0xf4
net/http.HandlerFunc.ServeHTTP(0xc000521640?, {0x1fc4330?, 0xc00006d500?}, 0xc0008b2b40?)
        /usr/local/go/src/net/http/server.go:2084 +0x2f
github.com/ansible-semaphore/semaphore/api.StoreMiddleware.func1.1()
        /go/src/github.com/ansible-semaphore/semaphore/api/router.go:26 +0x2e
github.com/ansible-semaphore/semaphore/db.StoreSession({0x1fc8518, 0xc0000b85f0}, {0xc0008b2b40, 0x1b}, 0xc000521698)
        /go/src/github.com/ansible-semaphore/semaphore/db/Store.go:336 +0x68
github.com/ansible-semaphore/semaphore/api.StoreMiddleware.func1({0x1fc4330, 0xc00006d500}, 0xc0008d2a00)
        /go/src/github.com/ansible-semaphore/semaphore/api/router.go:25 +0xfc
net/http.HandlerFunc.ServeHTTP(0xc0008d2a00?, {0x1fc4330?, 0xc00006d500?}, 0x975ea0?)
        /usr/local/go/src/net/http/server.go:2084 +0x2f
github.com/ansible-semaphore/semaphore/cli/cmd.runService.func1.1({0x1fc4330, 0xc00006d500}, 0xc000521778?)
        /go/src/github.com/ansible-semaphore/semaphore/cli/cmd/root.go:67 +0x124
net/http.HandlerFunc.ServeHTTP(0x7f970b53b3c8?, {0x1fc4330?, 0xc00006d500?}, 0xc000cc6cf0?)
        /usr/local/go/src/net/http/server.go:2084 +0x2f
github.com/gorilla/mux.CORSMethodMiddleware.func1.1({0x1fc4330, 0xc00006d500}, 0xc0002da3c0?)
        /go/src/github.com/ansible-semaphore/semaphore/vendor/github.com/gorilla/mux/middleware.go:51 +0xaa
net/http.HandlerFunc.ServeHTTP(0xc0008d2400?, {0x1fc4330?, 0xc00006d500?}, 0x9be0e0?)
        /usr/local/go/src/net/http/server.go:2084 +0x2f
github.com/gorilla/mux.(*Router).ServeHTTP(0xc0004ca000, {0x1fc4330, 0xc00006d500}, 0xc0008d2100)
        /go/src/github.com/ansible-semaphore/semaphore/vendor/github.com/gorilla/mux/mux.go:212 +0x202
github.com/gorilla/handlers.ProxyHeaders.func1({0x1fc4330, 0xc00006d500}, 0xc0008d2100)
        /go/src/github.com/ansible-semaphore/semaphore/vendor/github.com/gorilla/handlers/proxy_headers.go:59 +0x142
net/http.HandlerFunc.ServeHTTP(0x3b?, {0x1fc4330?, 0xc00006d500?}, 0x0?)
        /usr/local/go/src/net/http/server.go:2084 +0x2f
github.com/ansible-semaphore/semaphore/cli/cmd.cropTrailingSlashMiddleware.func1({0x1fc4330?, 0xc00006d500?}, 0xc0008d2100?)
        /go/src/github.com/ansible-semaphore/semaphore/cli/cmd/server.go:27 +0xf7
net/http.HandlerFunc.ServeHTTP(0x0?, {0x1fc4330?, 0xc00006d500?}, 0x40f025?)
        /usr/local/go/src/net/http/server.go:2084 +0x2f
net/http.serverHandler.ServeHTTP({0x1fc28c8?}, {0x1fc4330, 0xc00006d500}, 0xc0008d2100)
        /usr/local/go/src/net/http/server.go:2916 +0x43b
net/http.(*conn).serve(0xc00067a500, {0x1fc4708, 0xc000441ec0})
        /usr/local/go/src/net/http/server.go:1966 +0x5d7
created by net/http.(*Server).Serve
        /usr/local/go/src/net/http/server.go:3071 +0x4db
```

I'm not sure what I'm doing wrong, I'm merely copy-pasting the contents of the keypair's `id_rsa` file into the text box and hitting `Create`. The container is running behind an nginx proxy with this configuration:

```
server {
    server_name                     deploy.some.url.de;
    listen      443 ssl;
    listen [::]:443 ssl;

    ssl_certificate                 /etc/letsencrypt/live/some.url.de/fullchain.pem;
    ssl_certificate_key             /etc/letsencrypt/live/some.url.de/privkey.pem;

    location / {
        proxy_pass                  http://192.168.127.131:3000/;
        proxy_set_header Host $http_host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        
        proxy_set_header X-Forwarded-Proto $scheme;

        proxy_buffering off;
        proxy_request_buffering off;
    }

    location /api/ws {
        proxy_pass http://192.168.127.131:3000/api/ws;
        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "upgrade";
        proxy_set_header Origin "";
    }
}
```

But also when I port-forward port 3000 to my local machine via SSH and access it directly (thus bypassing nginx completely) I'm getting the same error.

Could you give me a hint on what the problem might be? Thanks 🙂

**Comment by kwanlowe** [comment]
I'm receiving a similar issue, though I suspect it's not the nginx configuration as I see the same behaviour when using the pure HTTP port.

```
                                                                                                                                                                                                                                               
time="2023-01-28T20:29:36Z" level=error msg="illegal base64 data at input byte 16"                                                                                                                                                             
goroutine 168 [running]:                                                                                                                                                                                                                       
runtime/debug.Stack()                                                                                                                                                                                                                          
        /usr/local/go/src/runtime/debug/stack.go:24 +0x65                                                                                                                                                                                      
runtime/debug.PrintStack()                                                                                                                                                                                                                     
        /usr/local/go/src/runtime/debug/stack.go:16 +0x19                                                                                                                                                                                      
github.com/ansible-semaphore/semaphore/api/helpers.WriteError({0x1fb54d0, 0xc00013a460}, {0x1fb25e0?, 0x22a87c0?})                                                                                                                             
        /go/src/github.com/ansible-semaphore/semaphore/api/helpers/helpers.go:92 +0x1ae                                                                                                                                                        
github.com/ansible-semaphore/semaphore/api/projects.AddKey({0x1fb54d0, 0xc00013a460}, 0xc0001452c0?)                                                                                                                                           
        /go/src/github.com/ansible-semaphore/semaphore/api/projects/keys.go:91 +0x66d        
```

I'm generating the SSH key with: ssh-keygen -b 2048 -t rsa
Trying to add a username/password also fails with similar errors in the log.
SELinux is disabled (did that after seeing the WriteError).

**Comment by kwanlowe** [comment]
I think I figured it out for my setup.  Make sure your access key is configured properly. There was some base64 conversion issue in my case. Once I fixed that, the base64 error went away.

**Comment by AlexDaichendt** [comment]
I was facing the same issue for both ssh keys and regular logins. Even when I clearly enter alphabetic values only. Apparently, this was related to my chosen encryption value: I generated an alphanumerical one myself, not with `head -c32 /dev/urandom | base64` like it was advised. Once I fixed that, it worked!

**Comment by rexdavid** [comment]
I am seeing the same issue with Docker image install.

**Comment by Svedrin** [comment]
> I was facing the same issue for both ssh keys and regular logins. Even when I clearly enter alphabetic values only. Apparently, this was related to my chosen encryption value: I generated an alphanumerical one myself, not with `head -c32 /dev/urandom | base64` like it was advised. Once I fixed that, it worked!

That was it. 🤦‍♂️ 

Knowing this, the error makes sense and the documentation states very clearly that the secret must be base64-encoded. Glad I'm not the only one not following the rules 😅 

I'm going to close this issue then. Hopefully it'll serve others as documentation.

**Comment by itfors** [comment]
> Hopefully it'll serve others as documentation.

So, "access_key_encryption" **shouldn't** be generated using `head -c32 /dev/urandom | base64` but rather by doing _what_ exactly?

...the docs still say:

- SEMAPHORE_ACCESS_KEY_ENCRYPTION — key for encrypting access keys in database. It must be generated by using the following command: head -c32 /dev/urandom | base64.

**Comment by Svedrin** [comment]
@itfors what the docs say is correct, and `head -c32 /dev/urandom | base64` will work. The problem was that I hadn't used it when setting up my Semaphore, but instead just typed some random characters which then weren't valid base64.

**Comment by itfors** [comment]
I've created a separate issue on the matter: #1339

**Comment by pwcazenave** [comment]
I had a similar issue. In my instance, I was starting the semaphore container via `systemd` and was passing in a value of `SEMAPHORE_ACCESS_KEY_ENCRYPTION` in an environment file (via `EnvironmentFile` in the systemd unit) as well as a mapped `config.json` (with `-v /var/lib/semaphore/config/config.json:/etc/semaphore/config.json:Z`). I removed the value from the environment file and left only a value in the config file and that resolved it. I also regenerated a new string based on #1339.
